### PR TITLE
Update requirements, restructure files and fix formatting for VAE example

### PR DESCRIPTION
**Pull request by canyon289**
# What does this PR do?

Addressing a portion of the issues in https://github.com/google/flax/issues/573
Putting up for review early to ensure I'm trending in the right direction and what else would be needed for a merge

Fixes
* File structure - copied from WMT Example
* Google style fixes - 99% done, some small errors may remain
* Updates Requirements 

I can fix up colab, readme, clu, model configs, and adding tests in separate PRs if you don't mind

### CODE REVIEW

**Comment by canyon289** [comment]
I see the linting issues. Will get them fixed

**Comment by 8bitmp3** [comment]
Thanks @canyon289 

@andsteing @cgarciae

File: examples/vae/input_pipeline.py
Comment (andsteing) [inline examples/vae/input_pipeline.py:41]: we usually create the `tfds.builder()` in the main `train.py` and pass it as an argument:
https://github.com/google/flax/blob/7c9b3b34a3aa325a5be63cd89b9d2ddf48407f79/examples/imagenet/train.py#L280

File: examples/vae/main.py
Comment (andsteing) [inline examples/vae/main.py:59]: nit: we usually have two empty lines before this block.

File: examples/vae/train.py
Comment (andsteing) [inline examples/vae/train.py:19]: nit: let's move this down to `import utils as vae_utils` to make it clear which are local imports.

File: examples/vae/train.py
Comment (andsteing) [inline examples/vae/train.py:23]: nit: let's move this down to `import utils as vae_utils` to make it clear which are local imports.

File: examples/vae/train.py
Comment (andsteing) [inline examples/vae/train.py:25]: nit insert empty space to separate local imports from other library imports.

File: examples/vae/input_pipeline.py
Comment (andsteing) [inline examples/vae/input_pipeline.py:44]: let's maybe remove the return type hints here (and in `build_test_set()` below) for now for better consistency

if you want you can add them everywhere in a second step, but in most examples we don't actually use typing annotations at all.

File: examples/vae/train.py
Comment (andsteing) [inline examples/vae/train.py:98]: when you create the datasetbuilder in this file (see my other comment in `input_pipeline.py`) then you can read the number of examples from `builder.info` instead of hard-coding it here.

File: examples/vae/input_pipeline.py
Comment (andsteing) [inline examples/vae/input_pipeline.py:44]: Let's pass in `batch_size` instead of `flags` here.

We usually like to have specific arguments, rather than passing around namespaces.

File: examples/vae/models.py
Comment (andsteing) [inline examples/vae/models.py:56]: also here, let's pass `latents` instead of `flags`

File: examples/vae/train.py
Comment (andsteing) [inline examples/vae/train.py:76]: also here, I would propose to pass in the individual values instead, to make the interface better defined.

**Review by andsteing — CHANGES_REQUESTED** [review]
This looks great, thanks!

Some minor comments inline.

**Comment by andsteing** [comment]
It's fine to clean-up README in a second step, but could you already update it to run `python main.py` instead of `python train.py` ?

**Comment by andsteing** [comment]
Tested that everything installs und runs (with expected loss) in a fresh Colab CPU runtime:
https://colab.research.google.com/drive/1MDrchufz1eUoB03znEnRV-Z8yQtg0Qau

File: examples/vae/input_pipeline.py
Comment (canyon289) [reply examples/vae/input_pipeline.py]: Moved, let me know if I put it in the right spot

File: examples/vae/main.py
Comment (canyon289) [reply examples/vae/main.py:64]: Fixed

File: examples/vae/train.py
Comment (canyon289) [reply examples/vae/train.py]: Fixed

File: examples/vae/train.py
Comment (canyon289) [reply examples/vae/train.py]: Also fixed

File: examples/vae/train.py
Comment (canyon289) [reply examples/vae/train.py:28]: Fixed

File: examples/vae/input_pipeline.py
Comment (canyon289) [reply examples/vae/input_pipeline.py]: i wont add type hints then! to be honest not my favorite thing to do and agree in examples its not as needed as the main ilbrary

File: examples/vae/train.py
Comment (canyon289) [inline examples/vae/train.py:123]: A question for understanding, should the train step be jitted? I looked at other examples and that didn't seem to be the case. 

https://github.com/google/flax/blob/main/examples/wmt/train.py#L166

**Comment by canyon289** [comment]
Do let me know if there's anything I missed. Happy to keep working on this

**Comment by 8bitmp3** [comment]
cc @andsteing @chiamp

**Comment by canyon289** [comment]
Im still committed to finishing this!

File: examples/vae/train.py
Comment (andsteing) [reply examples/vae/train.py:123]: it's jitted here:
https://github.com/google/flax/blob/df66c8128ede2a6b5642750f5c7c6e6e72605f1f/examples/wmt/train.py#L528

(the `pmap()` transform also compiles the code like `jit()`, but at the same time parallelizes it onto multiple devices)

Comment (andsteing) [reply examples/vae/train.py:123]: > should the train step be jitted?

to answer you question: yes, you should always compile the largest possble code block; usually that's the `train_step()`.

**Comment by andsteing** [comment]
@canyon289 if you would like reviewers to have another look at the PR, you can click on the `Re-quest review` button in the Github UI
image


(otherwise busy reviewers might not read all the individual updates and miss the PR until that button is pressed)

**Review by andsteing — APPROVED** [review]
Looks good from my side.
(one open comment about empty lines, otherwise good to submit)

**Comment by andsteing** [comment]
@cgarciae @8bitmp3 should we wait for your review?
(since you're both listed as reviewers)

**Comment by canyon289** [comment]
> What I meant: we usually have single empty lines within functions (lines 55-56, 59-60 should both be a single line), but we have double empty lines between "top-level codeblocks" (line 65 should be two empty lines)

You're right, updated! Thank you

File: examples/vae/train.py
Comment (canyon289) [reply examples/vae/train.py:123]: Ah I missed it that, was looking for the decorator. Thank you @andsteing

**Comment by canyon289** [comment]
> @canyon289 if you would like reviewers to have another look at the PR, you can click on the `Re-quest review` button in the Github UI image
> 
> (otherwise busy reviewers might not read all the individual updates and miss the PR until that button is pressed)

I dont have the rights in this repo to rerequest review from pending reviewers for some reason, I wouldnt mind another review if someone wants to, but dont want to obligate anybody. Happy to have this merged and keep moving on from there

**Comment by 8bitmp3** [comment]
@marcvanzee and @levskaya PTAL when you have time please

**Comment by andsteing** [comment]
> > @canyon289 if you would like reviewers to have another look at the PR, you can click on the `Re-quest review` button in the Github UI image
> > (otherwise busy reviewers might not read all the individual updates and miss the PR until that button is pressed)
> 
> I dont have the rights in this repo to rerequest review from pending reviewers for some reason, I wouldnt mind another review if someone wants to, but dont want to obligate anybody. Happy to have this merged and keep moving on from there

oh, that's good to know! I must have missed your message previously, since nobody else commented on the PR, let's move forward (I removed Cristian and Olzhas from the list of reiewers).

**Pull request merged**
